packages/phone-number-privacy/combiner/src/salt-generation/get-threshold-salt.ts

```diff
@@ -105,12 +105,100 @@ async function requestSignatures(request: Request, response: Response) {
 
   await Promise.all(signerReqs)
 
+  logResponseDiscrepancies(responses)
   const majorityErrorCode = getMajorityErrorCode(errorCodes)
-  logger.debug(
-    `Done requesting signatures. Success count: ${successCount}, majority error code ${majorityErrorCode}`
+  if (!blsCryptoClient.hasSufficientVerifiedSignatures()) {
+    handleMissingSignatures(majorityErrorCode, response)
+  }
+}
+
+async function handleSuccessResponse(
+  res: FetchResponse,
+  status: number,
+  response: Response,
+  sentResult: { sent: boolean },
+  responses: SignMsgRespWithStatus[],
+  serviceUrl: string,
+  blsCryptoClient: BLSCryptographyClient,
+  blindedQueryPhoneNumber: string
+) {
+  const signResponse = (await res.json()) as SignMessageResponse
+  if (!signResponse.success) {
+    // Continue on failure as long as signature is present to unblock user
+    const signResponseFailure = signResponse as SignMessageResponseFailure
```

Comment:
So this is exactly what I was trying to avoid, the pesky use of `x as y` which I consider "the death of passion" (direct translation from Romanian 😂).

I have to bring up @mcortesi 's point of erring towards union types with discriminator fields that help the compiler figure out object structure and allows us to avoid casting.

Here's an example of what I think the type definitions could look like: https://codesandbox.io/s/intelligent-banzai-ciqr1

This now takes into account your point that the actual body of the response could contain all fields in either the success or error case so there's a base interface that both cases extend, and both specific cases set the discriminator as the `success` value, and the failure case adds the `error` attribute. If there are other attributes that can be split they should be removed from the base interface.

Then the crux of the solution is the fact that we introduce a union type for the two interfaces. This is important because the compiler can then infer which branch of the union you are on based on the code. So once you've checked `if (!signResponse.success)` the compiler will automatically pick the right interface in each branch of the `if` statement and it will just flow naturally.

Reply:
There's also the one above `(await res.json()) as SignMessageResponse` is the one that can only be avoided by the runtime types discussed as well, but as we discussed we can live with it for now because of the overhead, but we should seriously think about enforcing that boundary as well.